### PR TITLE
Separate find files utility

### LINKED ISSUE #82: Separate file discovery from Scene init
There are currently two methods for initializing a Scene:

1. Provide the filenames for the files to be read as input. Optionally provide the name (or instance) of the reader to be used to open the provided files.
2. Provide information for sorting through available on-disk files in a `base_dir` and discover which reader should be used and which files match the parameters.

It has been determined from off-github conversation that the 2nd method should be separated out in to its own function. The keywords required to perform both methods being in the same `__init__` method has caused confusion among users, especially when compared to the legacy mpop interface.

Questions to answers:

1. Should the discovery function be available as a separate utility function or Scene classmethod? If utility function, where should it be located and should it be added to `satpy/__init__.py`? 
2. What keywords related to file and reader loading are presented in `Scene.__init__`?

    reader
    filenames
    reader_kwargs
    others?

3. Is `reader` still optional in the Scene init? Making it optional would allow the user to provide multiple formats of files and have multiple readers created to read those files. A feature that isn't heavily tested, but is intended to work.
4. Does the discovery function return a list of files or a dictionary of `reader_name: [filenames]`?
5. Do we allow for filtering based on metadata in the `Scene.__init__`? Does it have to be metadata only available inside the file or can it be metadata available from the filename? I believe filtering on filename metadata is currently available via the discovery functionality.
6. Does separating these interfaces stop us or make it more difficult to add certain features in the future?
7. Do we also change `**metadata` to just `metadata` or even remove it all together? Its function is to add metadata to the `Scene.info` dictionary.

### PR DESCRIPTION
**Pull request by davidh-ssec**
Fix #82 

TODO:

 - [x] Finalize Scene interface for this version and next
 - [x] Add at least one unit test of new utility function

### CODE REVIEW

File: satpy/readers/__init__.py [hunk 1]
```diff
@@ -249,18 +249,108 @@ def read_reader_config(config_files):
...
     reader_info = read_reader_config(reader_configs)
     reader_instance = reader_info['reader'](
         config_files=reader_configs,
-        metadata=metadata,
         **reader_kwargs
     )
     return reader_instance
 
 
+def configs_for_reader(reader=None, ppp_config_dir=None):
```
[the remark this answers is not in the document]
Comment (davidh-ssec) [reply satpy/readers/__init__.py]: Why is that a separate package?

File: satpy/composites/abi.py [hunk 2]
```diff
@@ -37,6 +37,15 @@ def four_element_average(d):
     return np.ma.mean(d.reshape(new_shape), axis=(1, 3))
 
 
+def simulated_green(c01, c02, c03):
```
[the remark this answers is not in the document]
Comment (davidh-ssec) [reply satpy/composites/abi.py]: No, but it was a small change so I thought I'd slip it in in this PR.

**Comment by davidh-ssec** [comment]
Current filtering cases that need to be handled consistently:

1. Filter filenames based on filename metadata including start and end time
2. Filter file handlers based on internal file (or data) start and end time (file handler properties)
3. Filter file handlers based on geographic area
4. Filter file handlers based on file metadata (`fh.metadata` dictionary)

So 2, 3, and 4 all require opening the files first. I believe @mraspaud and @pnuu and I discussed this on Slack, but to implement 2, 3, and 4 requires the user to do:

```
scn = Scene(..., reader_kwargs={'metadata': {'start_time': datetime(...)}})
```

Or similar. Is that too ugly? Because of how flexible this could be I don't think it makes sense to use `**kwargs` collection to the `Scene` and have that pass it along to the file handler filtering.

Another option for 2, 3, and 4 is to separate out the file handler filtering to a separate `Scene` method like `filter_data_sources` or something similar. The file handler filtering already has to create the complete file handler object so this isn't wasting processing time as far as that is considered.

Lastly, besides filtering by area, start time, and end time, is there really other metadata that someone would filter by that couldn't be taken from the filename? I suppose there is always that chance. Plus this does allow for some really fancy filtering later on. For example if the file handler wanted to do a "quality control" check on `__init__` the user could then filter out all bad files `scn.filter_data_sources(quality_threshold=0.8)`.

Thoughts @pnuu @mraspaud @adybbroe?

**Comment by davidh-ssec** [comment]
Additionally, I think the `Scene.__init__` parameters would eventually (after all deprecated features are removed) look like this:

```
def __init__(filenames, reader=None, ppp_config_dir=None, reader_kwargs=None):
```

**Comment by mraspaud** [comment]
I would rather have the file filtering from file data allowed within the __init__. Is `reader_kwargs` going to be used for more than filtering, or could we rename this to eg `filtering_parameters` ? That would make it clearer.

Regarding netcdf saving, what I do now is just `scn.save_datasets(write='cf', filename='bla.nc')`. You want to make a shortcut for that ?

**Comment by davidh-ssec** [comment]
The `reader_kwargs` could include other stuff and do in a few cases. My only example that I have right now is the NUCAPS reader where I have a `mask_surface` and a `mask_quality` keyword argument to do additional masking and filtering of the output data. I *think* these could be theoretically implemented as modifiers, but it's not how I've been thinking of it.

For NetCDF file, not necessarily a shortcut. The only difference (depending on how you implemented it) is that `scn.info` would be used as global attributes. I guess as long as `save_datasets` allows for global attributes in a keyword argument that there is no benefit to `scn.save(...)`.

**Comment by mraspaud** [comment]
Ok, so should we make the `filtering_parameters` separate for clarity ?

**Comment by davidh-ssec** [comment]
I'm ok with that since I'm not a huge fan of that double dict thing. Only odd gotcha with doing that would be if any of the filtering parameters should also be metadata of the reader...or I guess it just internally becomes metadata of the reader?

`Scene(..., filtering_parameters={...}) == Scene(..., reader_kwargs={metadata: {...}})`?

**Comment by davidh-ssec** [comment]
@mraspaud I started doing the `filter_parameters` changes (chose that instead of `filtering_parameters` and started to find some unnecessary code. First and maybe most importantly is that I removed the `sensor` keyword from the regular reader loading during `Scene.__init__`. This means that the `Scene` requires a list of filenames and the reader names used to read them. There is no check for `sensor` and there is no automatic reader finding. I also removed the `base_dir` keyword since it shouldn't be used anymore. This simplified the `ReaderFinder` class enough that I just turned it in to a function.

I started to consider moving all of the filename filtering stuff out of the reader base classes but then realized that it would require duplicating the config reading and stops subclasses from customizing any filename filtering.

Lastly, I'm wondering if we should remove the `start_time`, `end_time`, and `area` keywords as individual keywords to the readers and instead make them part of `filter_parameters` or other. They don't seem to be used (even by the custom `mipp_xrit` reader) for anything except filename or file handler filtering.

Although this is "aggressive" as far as removing features from the `Scene` init, I think it simplifies the init a ton so that new users are scared away by the number of options.

**Comment by pnuu** [comment]
Sounds good, although I haven't looked at the implementation. The only question arises from the `sensor`; are there any data formats that have several instruments in the same file? That would mean that the filename pattern can't be used to guess the reader. But then again, in those cases the reader should(?) be given explicitly by the user, right?.

**Comment by mraspaud** [comment]
Works for me. I like putting all relevant keywords to `filtering_parameters`, even `sensor`. Having filtering mixins is fine by me if it make the code more concise/simpler.

Just a reminder though that we need to provide an externally available function to fetch the filenames, so that the `base_dir` behaviour could be reproduced if needed by the user.

**Comment by davidh-ssec** [comment]
@pnuu Several sensors data in one file: Not that we support right now but I wouldn't be surprised if some exist. When you think of all the level 2 and higher level processing where collocation has to happen I could see people putting multiple in one.

@mraspaud Regarding `base_dir`, which functionality are you talking about?

**Comment by mraspaud** [comment]
I was just refering to having a drop in replacement for `Scene(base_dir=...)` like eg:
`Scene(find_files(base_dir=...), ...)`

File: satpy/readers/yaml_reader.py [hunk 3]
```diff
@@ -167,6 +160,7 @@ def select_files_from_directory(self, directory=None):
         for pattern in self.file_patterns:
             matching = glob.iglob(os.path.join(directory, globify(pattern)))
             filenames.extend(matching)
+            print(pattern, filenames)
```
Comment (mraspaud) [inline satpy/readers/yaml_reader.py]: Should there be a print statement here ?

File: satpy/scene.py [hunk 4]
```diff
@@ -44,63 +44,100 @@
...
-            reader: The name of the reader to use for loading the data.
-            base_dir (str): The directory to search for files containing the
+            filenames (iterable or dict): A sequence of files that will be used to load data from. A ``dict`` object
+                                          should map reader names to a list of filenames for that reader.
+            reader (str or list): The name of the reader to use for loading the data or a list of names.
+            filter_parameters (dict): Specify loaded file filtering parameters.
+                                      Shortcut for `reader_kwargs['filter_parameters']`.
+            reader_kwargs (dict): Keyword arguments to pass to specific reader instances.
+            ppp_config_dir (str): The directory containing the configuration files for satpy.
+            base_dir (str): (DEPRECATED) The directory to search for files containing the
```
Comment (mraspaud) [inline satpy/scene.py]: Should we have an alternative indicated in the DEPRECATED parentheses (like the others) ?

File: satpy/tests/test_readers.py [hunk 5]
```diff
@@ -117,68 +117,61 @@ def test_no_args(self):
...
+    #     """Test with default base_dir and reader specified"""
+    #     from satpy.readers import load_readers
+    #     fn = 'SVI01_npp_d20120225_t1801245_e1802487_b01708_c20120226002130255476_noaa_ops.h5'
+    #     # touch the file so it exists on disk
+    #     open(fn, 'w')
+    #     try:
+    #         ri = load_readers(reader='viirs_sdr')
+    #         self.assertListEqual(list(ri.keys()), ['viirs_sdr'])
+    #     finally:
+    #         os.remove(fn)
```
Comment (mraspaud) [inline satpy/tests/test_readers.py]: Why comments ?

File: satpy/tests/test_scene.py [hunk 6]
```diff
@@ -108,91 +96,68 @@ def test_init_with_ppp_config_dir(self):
...
-                                                   end_time=None,
-                                                   start_time=None)
-                findermock.return_value.assert_called_once_with(
-                    reader=reader_name,
-                    sensor=sensors,
-                    filenames=filenames,
-                    reader_kwargs=None,
-                    metadata={}
-                )
+    # TODO: Rewrite this test for the 'find_files_and_readers' function
```
Comment (mraspaud) [inline satpy/tests/test_scene.py]: Should this be done in this PR ?

File: satpy/readers/yaml_reader.py (hunk 3 above)
Comment (davidh-ssec) [reply satpy/readers/yaml_reader.py]: No... :(

File: satpy/scene.py (hunk 4 above)
Comment (davidh-ssec) [reply satpy/scene.py]: Yes.

File: satpy/tests/test_readers.py (hunk 5 above)
Comment (davidh-ssec) [reply satpy/tests/test_readers.py]: To remember that the tests used to exist and they should be reimplemented with the new interface.

File: satpy/tests/test_scene.py (hunk 6 above)
Comment (davidh-ssec) [reply satpy/tests/test_scene.py]: Yes...probably.

**Comment by davidh-ssec** [comment]
After I fix the conflicts, good to merge @mraspaud ? Maybe one more bug fix release with current develop then merge? I see that `mipp_xrit.py` is conflicting now which worries me because I thought we said that file isn't needed any more. Right?

**Comment by mraspaud** [comment]
Great job on the documentation, @davidh-ssec ! I can merge if you say the word.

**Pull request merged**